### PR TITLE
Refactor cosi interface

### PR DESCRIPTION
**Pull request by Gilthoniel**
This refactors the cosi interface so that it can verify the signatures during the protocol. The actual unit tests will come after this PR.

### CODE REVIEW

File: blockchain/skipchain/member.go [hunk 1]
```diff
@@ -48,25 +49,76 @@ func (c Conode) Pack() (proto.Message, error) {
 	return conode, nil
 }
 
+type iterator struct {
+	conodes []Conode
+	index   int
+}
+
+func (i *iterator) Next() bool {
```
Comment (nkcr) [inline blockchain/skipchain/member.go]: ~Next~ HasNext should only check `if i.index < len(i.conodes)` and ~Get~ GetNext should return the conode AND increment the index.

File: consensus/cosipbft/requests.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	req.digest, err = forwardLink.computeHash(f.New())
+	if err != nil {
+		return req, xerrors.Errorf("couldn't compute hash: %v", err)
+	}
+
+	return req, nil
+}
+
+// GetHash returns the hash of the proposal that will be signed by the collective
+// authority.
```
Comment (nkcr) [inline consensus/cosipbft/requests.go]:
```suggestion
// GetHash returns the hash of the prepare request that will be signed by the
// collective authority.
```

File: consensus/cosipbft/requests.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	commit := Commit{
+		to:      to,
+		prepare: prepare,
+		hash:    buffer,
+	}
+
+	return commit, nil
+}
+
+// GetHash returns the bytes of the prepare signature.
```
Comment (nkcr) [inline consensus/cosipbft/requests.go]: this comment is hard to understand: here we are returning the hash of the "request sent for the last phase of the PBFT". So `c.Hash` is actually the bytes of the prepare signature?

File: consensus/cosipbft/requests.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+func (c Commit) GetHash() []byte {
+	return c.hash
+}
+
+// Pack returns the protobuf message or an error,
```
Comment (nkcr) [inline consensus/cosipbft/requests.go]:
```suggestion
// Pack returns the protobuf message representation of a Commit, or an error
```

File: cosi/mod.go [hunk 5]
```diff
@@ -3,13 +3,22 @@ package cosi
...
+// PublicKeyIterator is an iterator over the list of public keys of a
+// collective authority.
+type PublicKeyIterator interface {
+	Next() bool
+	Get() crypto.PublicKey
+}
+
+// CollectiveAuthority (or Cothority in short) is a set of participant to a
+// collective signature.
+type CollectiveAuthority interface {
```
Comment (nkcr) [inline cosi/mod.go]: I appreciate the use of a non-abbreviated word 😇

File: mino/minoch/manager.go [hunk 6]
```diff
@@ -3,44 +3,55 @@ package minoch
...
+		return nil
+	}
+
+	return m.instances[string(text)]
 }
 
 func (m *Manager) insert(inst *Minoch) error {
 	addr := inst.GetAddress().(address)
-	if addr.String() == "" {
+	text, err := addr.MarshalText()
```
[the remark this answers is not in the document]
Comment (Gilthoniel) [reply mino/minoch/manager.go]: That's the standard golang interface for marshaling text.

File: blockchain/skipchain/block.go [hunk 7]
```diff
@@ -66,9 +68,14 @@ func (b SkipBlock) GetHash() []byte {
 	return b.hash[:]
 }
 
-// GetPublicKeys returns the list of public keys of the block.
-func (b SkipBlock) GetPublicKeys() []crypto.PublicKey {
-	return b.Conodes.GetPublicKeys()
+// GetPreviousHash returns the previous block digest.
+func (b SkipBlock) GetPreviousHash() []byte {
+	return b.BackLinks[0].Bytes()
```
[the remark this answers is not in the document]
Comment (Gilthoniel) [reply blockchain/skipchain/block.go]: A skipblock can only come from the factory from an external point of view and the factory makes sure that it is defined (among other checks).

[the remark this answers is not in the document]
Comment (Gilthoniel) [reply blockchain/skipchain/block.go]: (it will, not quite yet implemented)

**Pull request merged**